module/zfs/arc.c

```diff
@@ -4066,25 +4113,108 @@ arc_evict_state(arc_state_t *state, arc_buf_contents_t type, uint64_t spa,
 		multilist_sublist_unlock(mls);
 	}
 
+	evict_arg_t *evarg = kmem_alloc(sizeof (*evarg) * num_sublists,
+	    KM_SLEEP);
 	/*
 	 * While we haven't hit our target number of bytes to evict, or
 	 * we're evicting all available buffers.
 	 */
 	while (total_evicted < bytes) {
 		int sublist_idx = multilist_get_random_index(ml);
+		boolean_t usetskq = zfs_arc_evict_parallel;
 		uint64_t scan_evicted = 0;
 
+		uint64_t left = (bytes == ARC_EVICT_ALL ? bytes :
+		    bytes - total_evicted);
+
+		/*
+		 *  How we scale
+		 *
+		 *  Example 1, # of chunks less than # of tasks.
+		 *  We have:
+		 *  - 4 tasks
+		 *  - 3 chunks
+		 *  - 3 full col
+		 *  - 0 low cols.
+		 *
+		 *  The first low col index is 3.
+		 *  The tasks #0-#2 evict 1 chunk each.
+		 *
+		 *    0 | 1 | 2 | 3 |
+		 *  +===+===+===+===+
+		 *  | x | x | x |   |
+		 *  +---+---+---+---+
+		 *
+		 *  Example 2, # of chunks more than # of tasks.
+		 *  We have:
+		 *  - 4 tasks
+		 *  - 9 chunks
+		 *  - 1 full col
+		 *  - 3 low cols
+		 *
+		 *  The first low col index is 1.
+		 *  The task #0 evicts 3 chunks, the others evict 2 chunks each.
+		 *
+		 *    0 | 1 | 2 | 3 |
+		 *  +===+===+===+===+
+		 *  | x | x | x | x |
+		 *  +---+---+---+---+
+		 *  | x | x | x | x |
+		 *  +---+---+---+---+
+		 *  | x |   |   |   |
+		 *  +---+---+---+---+
+		 */
+
+		/*
+		 * Compute number of tasks to run (n), low col index (k)
+		 * and normal and low bytes per task.
+		 */
+		uint64_t nchunks = ((left - 1) >> MIN_EVICT_PERTASK_SHIFT) + 1;
+		unsigned n = nchunks < num_sublists ? nchunks : num_sublists;
+		uint64_t fullrows = nchunks / n;
+		unsigned lastrowcols = nchunks % n;
+		unsigned k = (lastrowcols ? lastrowcols : n);
+
+		uint64_t bytes_pertask_low =
+		    fullrows << MIN_EVICT_PERTASK_SHIFT;
+		uint64_t bytes_pertask = bytes_pertask_low + (lastrowcols ?
+		    (1 << MIN_EVICT_PERTASK_SHIFT) : 0);
```

Comment:
I think you are over-engineering here.  I don't think eviction per taskq should really be a multiple of `1 << MIN_EVICT_PERTASK_SHIFT` to complicate the logic, merely it should be bigger than one.  So you could just use `MIN_EVICT_PERTASK_SHIFT` to decide number of tasks, and then split the eviction amount equally between them.

And I wonder if it would make sense to scale number of tasks with eviction not linearly, but in some logarithimic fashion to not spin too many threads at once, stressing the system more for diminishing return.

Reply:
This has nothing to do with taskqs.  It simply makes no sense to create multiple parallel jobs/tasks and wake up multiple threads/CPUs for operations below certain size due to overheads.  `SPA_MAXBLOCKSHIFT` is an absolute minimum there, since it may be impossible to free less than that any way if large blocks are used, but something bigger could be chosen based on some practical tests.  So just, as I have told, remove all this unneeded complexity of trying to be multiple to `MIN_EVICT_PERTASK_SHIFT`, select number of tasks based on total size (divide total size by the minimum size, rounding up, and limit the result by number of threads), and then give each task its fraction of the total size (just divide total size by number of tasks, rounding up).